File: okCore/src/main/java/unics/okcore/converter/moshi/MoshiConverter.java
package unics.okcore.converter.moshi;

import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;
import com.squareup.moshi.Types;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.List;

import unics.okcore.converter.JsonConverter;

/**
 * Created by Lucio on 2022/1/12.
 */
public class MoshiConverter implements JsonConverter {

    private final Moshi moshi;

    public MoshiConverter(Moshi moshi) {
        this.moshi = moshi;
    }

    public MoshiConverter() {
        this(createPreferredMoshi());
    }

    @Nullable
    @Override
    public <T> T toObject(@Nullable String json, @NotNull Class<T> clazz) {
        if (json == null || json.isEmpty()) return null;
        try {
            return moshi.adapter(clazz).fromJson(json);
        } catch (IOException e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
    }

    @Nullable
    @Override
    public <T> T toObject(@Nullable String json, @NotNull Type type) {
        if (json == null || json.isEmpty()) return null;
        try {
            JsonAdapter<T> ja = moshi.adapter(type);
            return ja.fromJson(json);
        } catch (IOException e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
    }

    @Override
    public <T> List<T> toObjectList(@Nullable String json, @NotNull Class<T> tClass) {
        if (json == null || json.isEmpty()) return null;
        try {
            Type type = Types.newParameterizedType(List.class, tClass);
            JsonAdapter<List<T>> ja = moshi.adapter(type);
            return ja.fromJson(json);
        } catch (IOException e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
    }

    @Override
    public <T> String toJson(T obj, @NotNull Class<T> tClass) {
        return moshi.adapter(tClass).toJson(obj);
    }

    @Override
    public <T> String toJson(T obj, @NotNull Type type) {
        return moshi.adapter(type).toJson(obj);
    }

    public static Moshi createPreferredMoshi() {
        return new Moshi.Builder().build();
    }

    public static Moshi.Builder applyUtcDate(Moshi.Builder builder, boolean isSupportTimestamp) {
        builder.add(new UTCDateJsonAdapter(isSupportTimestamp));
        return builder;
    }
}

File: okCore/src/main/java/libcore/converter/moshi/MoshiConverter.kt
package libcore.converter.moshi

import libcore.converter.JsonConverter
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import java.io.IOException
import java.lang.reflect.Type

/**
 * Created by Lucio on 2022/1/12.
 */
class MoshiConverter @JvmOverloads constructor(private val moshi: Moshi = createPreferredMoshi()) :
    JsonConverter {

    override fun <T> toObject(json: String?, clazz: Class<T>): T? {
        return try {
            if (json.isNullOrEmpty()) return null
            moshi.adapter(clazz).fromJson(json)
        } catch (e: IOException) {
            e.printStackTrace()
            throw RuntimeException(e)
        }
    }

    override fun <T : Any?> toObject(json: String?, type: Type): T? {
        if (json.isNullOrEmpty()) return null
        return try {
            moshi.adapter<T>(type).fromJson(json)
        } catch (e: IOException) {
            e.printStackTrace()
            throw RuntimeException(e)
        }
    }

    override fun <T> toObjectList(json: String?, tClass: Class<T>): List<T>? {
        if (json.isNullOrEmpty()) return emptyList()
        return try {
            val type: Type = Types.newParameterizedType(MutableList::class.java, tClass)
            moshi.adapter<List<T>>(type).fromJson(json)
        } catch (e: IOException) {
            e.printStackTrace()
            throw RuntimeException(e)
        }
    }

    override fun <T : Any?> toJson(obj: T?, tClass: Class<T>): String? {
        return moshi.adapter(tClass).toJson(obj)
    }

    override fun <T : Any?> toJson(obj: T?, type: Type): String? {
        return moshi.adapter<T>(type).toJson(obj)
    }

    companion object {
        //    @Override
        //    public @Nullable String toJson(@Nullable Object obj) {
        //        return moshi.adapter(obj.getClass()).toJson(obj);
        //    }
        @JvmStatic
        fun createPreferredMoshi(): Moshi {
            return Moshi.Builder().build()
        }

        /**
         * @param isSupportTimestamp 是否支持时间戳
         */
        @JvmStatic
        fun applyUtcDate(
            builder: Moshi.Builder,
            isSupportTimestamp: Boolean = false
        ): Moshi.Builder {
            builder.add(UTCDateJsonAdapter(isSupportTimestamp))
            return builder
        }
    }
}

//
////
////    //反序列化
//inline fun <reified T> toJson2(t: T) = getAdapter<T>().toJson(t) ?: ""
//
//fun <T> getAdapter(type: Type): JsonAdapter<T> =
//    (Converters.getJsonConverter() as MoshiConverter).moshi.adapter(type)
//
//inline fun <reified T> getAdapter(): JsonAdapter<T> =
//    (Converters.getJsonConverter() as MoshiConverter).moshi.adapter(object : TypeToken<T>() {}.type)
//
//
//abstract class TypeToken<T> {
//    val type: Type
//        get() = run {
//            val superclass = javaClass.genericSuperclass
//            Util.canonicalize((superclass as ParameterizedType).actualTypeArguments[0])
//        }
//}